src/rhsmlib/dbus/util.py

```diff
@@ -21,22 +21,93 @@
 import re
 
 from rhsmlib.dbus import exceptions
+from rhsmlib.dbus import dbus_utils
+from rhsmlib.client_info import DBusSender
 
 log = logging.getLogger(__name__)
 
 __all__ = [
+    'dbus_get_sender_cmd_line',
+    'dbus_set_sender_cmd_line',
     'dbus_handle_exceptions',
+    'dbus_handle_sender',
     'dbus_service_method',
     'dbus_service_signal'
 ]
 
 
+def dbus_get_sender_cmd_line(sender, bus=None):
+    """
+    Try to get command line of sender
+    :param sender: sender
+    :param bus: bus
+    :return:
+    """
+    if bus is None:
+        bus = dbus.SystemBus()
+    cmd_line = dbus_utils.command_of_sender(bus, sender)
+    if cmd_line is not None and type(cmd_line) == str:
+        # Store only first argument of command line (no argument including username or password)
+        cmd_line = cmd_line.split()[0]
+    return cmd_line
+
+
+def dbus_set_sender_cmd_line(sender, cmd_line=None, bus=None):
+    """
+    This method set sender's command line in the singleton object
+    :return: None
+    """
+    dbus_sender = DBusSender()
+    if cmd_line is None:
+        dbus_sender.cmd_line = dbus_get_sender_cmd_line(sender, bus)
+    else:
+        dbus_sender.cmd_line = cmd_line
+    log.debug("D-Bus sender: %s (cmd-line: %s)" % (sender, dbus_sender.cmd_line))
+
+
+def dbus_reset_sender_cmd_line():
+    """
+    Reset sender's command line
+    :return: None
+    """
+    dbus_sender = DBusSender()
+    dbus_sender.cmd_line = None
+
+
+@decorator.decorator
+def dbus_handle_sender(func, *args, **kwargs):
+    """
+    Decorator to handle sender argument
+    :param func: method with implementation of own logic of D-Bus method
+    :param args: arguments of D-Bus method
+    :param kwargs: keyed arguments of D-Bus method
+    :return: result of D-Bus method
+    """
+
+    sender = None
+    # Get sender from arguments
+    if 'sender' in kwargs:
+        sender = kwargs['sender']
+    elif len(args) > 0:
+        sender = args[-1]
+
+    if sender is not None:
+        dbus_set_sender_cmd_line(sender)
+
+    try:
+        return func(*args, **kwargs)
+    finally:
+        dbus_reset_sender_cmd_line()
```

Comment:
And if we acquire the lock before the function call, I'd suggest we release it here.

Admittedly this is effectively serializing all our dbus calls which may be worth more discussion.

Reply:
I implemented this using `__enter__()` and `__exit__()` methods and `with DBusSender as dbus_sender` statement. It is more clear solution.